gcs/constraints/smart_table.cc

```diff
@@ -0,0 +1,642 @@
+        propagators.define_pseudoboolean_ge(state, {{1_i, flag_lt}, {1_i, flag_gt}}, 1_i, flag);
+        return flag;
+    case GREATER_THAN_EQUAL:
+        flag = propagators.create_proof_flag("bin_gt");
+        propagators.define_pseudoboolean_ge(state, move(var_1_minus_var_2), 0_i, flag);
+        propagators.define_pseudoboolean_ge(state, move(var_2_minus_var_1), 1_i, ! flag);
+        return flag;
+    default:
+        throw UnexpectedException{"Unexpected SmartEntry type encountered while creating PB model."};
+    }
+}
+
+auto literal_from_unary_entry(UnaryValueEntry & unary_entry) -> Literal
+{
+    auto var = unary_entry.var;
+    auto value = unary_entry.value;
+    switch (unary_entry.constraint_type) {
+    case LESS_THAN:
+        return var < value;
+    case LESS_THAN_EQUAL:
+        return var < value + 1_i;
+    case EQUAL:
+        return var == value;
+    case NOT_EQUAL:
+        return var != value;
+    case GREATER_THAN:
+        return var >= value + 1_i;
+    case GREATER_THAN_EQUAL:
+        return var >= value;
+    default:
+        throw UnexpectedException{"Unexpected SmartEntry type encountered while creating PB model."};
+    }
+}
+
+auto SmartTable::clone() const -> unique_ptr<Constraint>
+{
+    return make_unique<SmartTable>(_vars, _tuples);
+}
+
+auto SmartTable::install(Propagators & propagators, State & initial_state) && -> void
+{
+    vector<IntegerVariableID> selectors;
+
+    for (unsigned int i = 0; i < _tuples.size(); ++i) {
+        selectors.emplace_back(initial_state.allocate_integer_variable_with_state(0_i, 1_i));
+    }
+
+    if (propagators.want_definitions()) {
+        vector<ProofFlag> pb_selectors;
+
+        for (unsigned int i = 0; i < _tuples.size(); ++i) {
+            pb_selectors.emplace_back(propagators.create_proof_flag("t" + std::to_string(i)));
+        }
+        WeightedPseudoBooleanTerms sum_pb_selectors{};
+
+        for (const auto & s : pb_selectors) {
+            WeightedPseudoBooleanTerm t{1_i, s};
+            sum_pb_selectors.emplace_back(t); // No idea why, but this errors if I put the initialiser list directly in emplace_back()
+        }
+
+        propagators.define_pseudoboolean_ge(initial_state, move(sum_pb_selectors), 1_i);
+
+        // Would need a hash function for unordered map, but this shouldn't be too slow
+        std::map<BinaryEntryData, ProofFlag> smart_entry_flags;
+
+        for (unsigned int tuple_idx = 0; tuple_idx < _tuples.size(); ++tuple_idx) {
+            WeightedPseudoBooleanTerms entry_flags_sum{};
+            WeightedPseudoBooleanTerms entry_flags_neg_sum{};
+            for (const auto & entry : _tuples[tuple_idx]) {
+                overloaded{
+                    [&](BinaryEntry binary_entry) {
+                        auto binary_entry_data = std::make_tuple(binary_entry.var_1, binary_entry.var_2, binary_entry.constraint_type);
+                        if (! smart_entry_flags.contains(binary_entry_data))
```

Comment:
I think modern C++ has a function that does these two lines a bit more efficiently, `try_emplace` or something like that?